Gh 639 policy controller

maleck13 opened the pull request:
Draft PR adding the new policy controller to the deployment and CSV for kuadrant operator

This PR adds in the policy controller via kustomize it also add metal lb as an option

**Verification**
`make local-cluster-setup`
Build a new bundle and catalog and install it in to the cluster

expect to see a new `kuadrant-operator-policy-controller` in the `kuadrant-system` namespace
validate that the DNSPolicy and TLSPolicy CRDS are present in the cluster

**Extra Validation**

Before doing this you will need an AWS credential that gives access to route53 a domain (talk to craig as he can set you up with this)

deploy metallb to a local cluster
`make install-metallb`
`k create -f config/metallb/metal-lb.yaml`

create a new gateway isito gateway

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: prod-web
spec:
  gatewayClassName: istio
  listeners:
  - allowedRoutes:
      namespaces:
        from: All
    name: specific
    hostname: 'REPLACEME'
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        name: REPLACEME
  ```

Setup TLS

create an issuer

```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: glbc-ca
spec:
  selfSigned: {}

```

create a TLSPolicy

```
apiVersion: kuadrant.io/v1alpha1 
kind: TLSPolicy 
metadata: 
  name: prod-web
spec: 
  targetRef: 
    name: prod-web 
    group: gateway.networking.k8s.io 
    kind: Gateway    
  issuerRef: 
    group: cert-manager.io
    kind: ClusterIssuer
    name: glbc-ca
```

Observe the gateway is now configured and has an address and tls secrets

```
get gateway
NAME       CLASS   ADDRESS        PROGRAMMED   AGE
prod-web   istio   172.31.200.0   True         7m56s
```

Setup DNS
setup a dnsprovider 
https://docs.kuadrant.io/multicluster-gateway-controller/docs/dnspolicy/dns-provider/
setup a managedzone
https://docs.kuadrant.io/multicluster-gateway-controller/docs/dnspolicy/managed-zone/

create a dnspolicy

```
apiVersion: kuadrant.io/v1alpha1
kind: DNSPolicy
metadata:
  name: prod-web
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: prod-web
  routingStrategy: simple  
```

setup a httproute and backend

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: echo2
spec:
  parentRefs:
  - kind: Gateway
    name: prod-web
  hostnames:
  - specific.cb.hcpapps.net
  rules:
  - backendRefs:
    - name: echo
      port: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: echo
spec:
  ports:
    - name: http-port
      port: 8080
      targetPort: http-port
      protocol: TCP
  selector:
    app: echo     
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: echo
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
        - name: echo
          image: docker.io/jmalloc/echo-server
          ports:
            - name: http-port
              containerPort: 8080
              protocol: TCP 

```

should be able to dig your domain or curl it if you can resolve the IP

example `dig specific.cb.hcpapps.net`

clean up

```
k delete dnspolicy prod-web
k delete tlspolicy prod-web
```

maleck13 commented:
hmm code coverage went down but only changed YAMLs

didierofrivia reviewed (approved):
It works as expected, creating the DNS records and the TLS connection. I've done so setting up AWS as my DNS provider and creating a ManagedZone for my custom host, then applying the DNSPolicy. I've setup everything in 'default' namespace, leaving
"kuadrant-system for the operator/controllers work. You'll need to rebase and fix the conflicts of the bundle

didierofrivia commented on Makefile:
```diff
@@ -368,6 +368,14 @@ deploy-dependencies: kustomize dependencies-manifests ## Deploy dependencies to
 	kubectl -n "$(KUADRANT_NAMESPACE)" wait --timeout=300s --for=condition=Available deployments --all
 
+.PHONY: install-metallb
+install-metallb: $(KUSTOMIZE)
```
minor:

you might want to add a description next to the target in order to appear in the help command

```suggestion
install-metallb: $(KUSTOMIZE) ## Installs MetalLB Load Balancer.
```

maleck13 replied:
done

maleck13 commented:
@didierofrivia I have updated with a comment and rebased.

@alexsnaps thoughts on merging this now? We don't have any docs for DNS and TLS policy in the single cluster context yet, so this would put the policy controller in place as part of the release for gwapi v1 but would have no supporting docs right now. A "stealth" deploy if you will. Docs will come soon and def part of our next "unified kuadrant" release

guicassolato commented on config/metallb/metal-lb.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: metallb.io/v1beta1
```
Maybe put it in `./examples`?

maleck13 replied:
sure done

guicassolato replied:
I see you've opted to remove the file altogether.

guicassolato reviewed (approved):
All the steps I had to follow in my environment to make this work completely, until getting a 200 response to the curl command:

```sh
pwd
# ~/go/src/github.com/kuadrant/kuadrant-operator
```

```sh
export KIND_EXPERIMENTAL_DOCKER_NETWORK=kuadrant-local

docker network create -d bridge --subnet 172.31.0.0/16 $KIND_EXPERIMENTAL_DOCKER_NETWORK --gateway 172.31.0.1 \
  -o "com.docker.network.bridge.enable_ip_masquerade"="true" \
  -o "com.docker.network.driver.mtu"="1500"

make local-cluster-setup
```

```sh
make install-olm 
make generate manifests bundle
make deploy-catalog CATALOG_IMG=quay.io/kuadrant/kuadrant-operator-catalog:gh-639-policy-controller
```

```sh
kubectl get crd/dnspolicies.kuadrant.io
# NAME                      CREATED AT
# dnspolicies.kuadrant.io   2023-11-16T12:54:04Z
```

```sh
kubectl get crd/tlspolicies.kuadrant.io
# NAME                      CREATED AT
# tlspolicies.kuadrant.io   2023-11-16T12:54:04Z
```

```sh
export TMP_DIR=/tmp/kuadrant-local
mkdir -p ${TMP_DIR}/kubeconfigs
bin/kind export kubeconfig --name kuadrant-local --kubeconfig ${TMP_DIR}/kubeconfig --internal
bin/kind get kubeconfig --name kuadrant-local > ${TMP_DIR}/kubeconfigs/kuadrant-local.kubeconfig
bin/kind export kubeconfig --name kuadrant-local --kubeconfig ${TMP_DIR}/kubeconfigs/internal/kuadrant-local.kubeconfig --internal
mkdir -p ${TMP_DIR}/kubeconfigs/external/
EXTERNAL_KUBECONFIG=${TMP_DIR}/kubeconfigs/external/kuadrant-local.kubeconfig
cp ${TMP_DIR}/kubeconfigs/kuadrant-local.kubeconfig ${EXTERNAL_KUBECONFIG}
master_ip=$(docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' "kuadrant-local-control-plane" | head -n 1)
yq -i ".clusters[0].cluster.server = \"https://${master_ip}:6443\"" "${EXTERNAL_KUBECONFIG}"; sed -b 's/\x1b[^m]*m//g' -i "${EXTERNAL_KUBECONFIG}"
yq -i "(.. | select(. == \"kind-kuadrant-local\")) = \"kuadrant-local\"" "${EXTERNAL_KUBECONFIG}"; sed -b 's/\x1b[^m]*m//g' -i "${EXTERNAL_KUBECONFIG}"
chmod a+r "${EXTERNAL_KUBECONFIG}"
```

```sh
make install-metallb
```

```sh
kubectl -n metallb-system apply -f -<<EOF
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: kuadrant-local
spec:
  addresses:
  - 172.31.200.0/24
---
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: empty
EOF
```

```sh
export ROOT_DOMAIN=<ROOT_DOMAIN>
```

```sh
kubectl create namespace ingress
```

```sh
kubectl -n ingress apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: prod-web
spec:
  gatewayClassName: istio
  listeners:
  - allowedRoutes:
      namespaces:
        from: All
    name: specific
    hostname: "*.$ROOT_DOMAIN"
    port: 443
    protocol: HTTPS
    tls:
      mode: Terminate
      certificateRefs:
      - kind: Secret
        name: gui-hcpapps-tls
EOF
```

```sh
kubectl -n ingress apply -f -<<EOF
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: prod-ca
spec:
  selfSigned: {}
EOF
```

```sh
kubectl -n ingress apply -f -<<EOF
apiVersion: kuadrant.io/v1alpha1 
kind: TLSPolicy 
metadata: 
  name: prod-web
spec: 
  targetRef: 
    name: prod-web 
    group: gateway.networking.k8s.io 
    kind: Gateway    
  issuerRef: 
    group: cert-manager.io
    kind: ClusterIssuer
    name: prod-ca
EOF
```

```sh
export AWS_ACCESS_KEY_ID=… \
       AWS_REGION=… \
       AWS_SECRET_ACCESS_KEY=… \
       AWS_DNS_PUBLIC_ZONE_ID=…
```

```sh
kubectl -n ingress create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_REGION=$AWS_REGION \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
```

```sh
kubectl -n ingress apply -f - <<EOF
apiVersion: kuadrant.io/v1alpha1
kind: ManagedZone
metadata:
  name: prod-cluster
spec:
  id: $AWS_DNS_PUBLIC_ZONE_ID
  domainName: $ROOT_DOMAIN
  description: Kuadrant single cluster
  dnsProviderSecretRef:
    name: aws-credentials
    namespace: ingress
EOF
```

```sh
kubectl -n ingress apply -f -<<EOF
apiVersion: kuadrant.io/v1alpha1
kind: DNSPolicy
metadata:
  name: prod-web
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: prod-web
  routingStrategy: simple
EOF
```

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: echo
spec:
  parentRefs:
  - kind: Gateway
    name: prod-web
    namespace: ingress
  hostnames:
  - echo.$ROOT_DOMAIN
  rules:
  - backendRefs:
    - name: echo
      port: 8080
---
apiVersion: v1
kind: Service
metadata:
  name: echo
spec:
  ports:
  - name: http-port
    port: 8080
    targetPort: http-port
    protocol: TCP
  selector:
    app: echo     
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: echo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: echo
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
      - name: echo
        image: docker.io/jmalloc/echo-server
        ports:
        - name: http-port
          containerPort: 8080
          protocol: TCP 
EOF
```

```sh
dig echo.$ROOT_DOMAIN
# […]
# ;; ANSWER SECTION:
# echo.<ROOT_DOMAIN>.	60	IN	A	172.31.200.0
```

```sh
curl -k https://echo.$ROOT_DOMAIN -i
# HTTP/2 200
```

I could probably simplify the part of setting up the docker network, but it was inspired by https://github.com/Kuadrant/multicluster-gateway-controller/blob/f93bb5554ef813e9d85780774f6da0f3e57f2826/hack/quickstart-setup.sh#L85

maleck13 commented:
Nice, Thanks for the detailed steps. Could be useful to add these to a script or doc I guess for setting up Kuadrant single cluster
